routingtable/client_manager.go

```diff
@@ -1,21 +1,41 @@
 package routingtable
 
+// ClientManager manages clients of routing tables (observer pattern)
 type ClientManager struct {
-	clients      map[RouteTableClient]struct{} // Ensures a client registers at most once
-	routingTable *RoutingTable
+	clients map[RouteTableClient]struct{} // Ensures a client registers at most once
+	master  RouteTableClient
 }
 
+// NewClientManager creates and initializes a new client manager
+func NewClientManager(master RouteTableClient) ClientManager {
+	return ClientManager{
+		clients: make(map[RouteTableClient]struct{}, 0),
+		master:  master,
+	}
+}
+
+// Register registers a client for updates
 func (c *ClientManager) Register(client RouteTableClient) {
 	if c.clients == nil {
 		c.clients = make(map[RouteTableClient]struct{}, 0)
 	}
 	c.clients[client] = struct{}{}
-	//c.routingTable.updateNewClient(client)
+	c.master.UpdateNewClient(client)
 }
 
+// Unregister unregisters a client
 func (c *ClientManager) Unregister(client RouteTableClient) {
 	if _, ok := c.clients[client]; !ok {
 		return
 	}
 	delete(c.clients, client)
 }
+
+// Clients returns a list of registered clients
+func (c *ClientManager) Clients() (ret []RouteTableClient) {
+	for rtc := range c.clients {
+		ret = append(ret, rtc)
+	}
+
+	return ret
```

Comment:
i would prefer Clients() returning an empty slice instead of nil when no client is registered.

Reply:
ack. Fixed.